crates/fluvio-protocol-derive/src/ast/prop.rs

```diff
@@ -151,76 +143,128 @@ impl UnnamedProp {
         }
     }
 }
-
-pub fn validate_versions(min: i16, max: Option<i16>, field: Option<&str>) -> Option<String> {
```

Comment:
This place was used to validate on compile time. Because min_version and max_version support constants and functions they can only be evaluated at run time, so we can't actually use the validate_versions as it was before, that's why I removed it. 

There is validation for the runtime which check the token stream directly. 

https://github.com/infinyon/fluvio/blob/4b9b32db623f39af18a0a8f95c75d66277d1e786/crates/fluvio-protocol-derive/src/ast/prop.rs#L47-L63

Even if we change that to throw errors with to_compile_error in the token stream it will not work again because the TokenStream that loads the values for min/max versions won't be evaluated.

Reply:
Ok, let's do it in runtime then. If users mess up with versions, at least they will receive a proper message during encoding/decoding, it's better than nothing.

Reply:
We already have the runtime checks, so I haven't added anything more to it.

Reply:
What runtime check do you mean? We used to check that min is not greater than max, min is not negative, etc. We don't have it now, don't we?

Reply:
There are checks here that check the version.

https://github.com/infinyon/fluvio/blob/4b9b32db623f39af18a0a8f95c75d66277d1e786/crates/fluvio-protocol-derive/src/ast/prop.rs#L47-L63

Reply:
Previously the `validate_versions `only did checks and threw error of something was returned. We cannot throw syn::Error in this case. We need to have a code generation in our case in order to have the values evaluated before using them and then we throw an error. Not sure how you want the error to be thrown but currently the old code as I see prints the data with tracing is that not enough?

Reply:
I made it so we make checks with the generated code and panic if it finds an issue. I used the original checks and errors as text. Does that look okay?

Reply:
I found a way of doing it in compile time for literals, consts and const fn. `assert` is stable in const context already. We can leverage it. See the example here https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a58bfe08d0bb0ea8f579e488d124b728

We can drop support for non-const function for now.

Reply:
That is a pretty cool hack doing it like this: `const _: ()` . Currently I replaced all the panics with the const asserts and pretty much didn't change anything else, because even if you try to use a normal function it will still throw an error so no need to handle that differently than how it works.